### PR TITLE
Rework to use an abstract base class.

### PR DESCRIPTION
**Pull request by dpassen**
- Introduce separate (private) classes, _Something and _Nothing
  that inherit from _Optional.
- Introduce two top level functions, of and empty to simplify
  usage.
- Remove the tracking of check before get. Use of get should
  be heavily discouraged.

### CODE REVIEW

File: optional/Optional.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
+from abc import abstractmethod
+from .compatible_abc import CompatibleABC
+
+
+def of(thing):
```
Comment (cbefus) [inline optional/Optional.py]: this feels a bit odd to me.  I get that using the capital "O" Optional for filename it reads like you are doing:
```
Optional.of("Foo")
```
but that file naming is unconventional.  I still wonder about putting the static constructors into the Optional base class?

File: optional/Optional.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        return procedure()
+
+    def map(self, func):
+        return self
+
+    def flat_map(self, func):
+        return self
+
+    def __eq__(self, other):
+        return isinstance(other, _Nothing)
```
Comment (cbefus) [inline optional/Optional.py]: nice!

File: optional/compatible_abc.py [hunk 3]
```diff
@@ -0,0 +1,4 @@
+from abc import ABCMeta
+
+
+CompatibleABC = ABCMeta('ABC', (object,), {'__slots__': ()})
```
Comment (cbefus) [inline optional/compatible_abc.py]: I think we need an inline comment here `# for python 2 and 3 compatibility`

Comment (dpassen) [reply optional/compatible_abc.py]: I thought moving it out into its own file did enough. =\

File: optional/Optional.py (hunk 1 above)
Comment (dpassen) [reply optional/Optional.py]: I don't want them inherited by the derived classes. I tried a little __init__.py magic

File: optional/Optional.py (hunk 2 above)
Comment (dpassen) [reply optional/Optional.py]: Yeah

**Comment by dpassen** [comment]
Don't merge this either way. I'll want to squash

File: optional/Optional.py (hunk 1 above)
Comment (dpassen) [reply optional/Optional.py]: I ended up making a class called Optional with just the two static methods. No weird filenames, nothing weird in __init__.py

File: optional/compatible_abc.py (hunk 3 above)
Comment (dpassen) [reply optional/compatible_abc.py]: I added a package docstring.

File: optional/optional.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
+from abc import abstractmethod
+
+from .compatible_abc import CompatibleABC
+
+
+class Optional(object):
```
Comment (cbefus) [inline optional/optional.py]: nice!

**Pull request merged**